File: kotlin-in-action-master/src/ch03/study/study_3_2/collection.java
package ch03.study.study_3_2;

import java.util.*;

import Maths.MathFunctions;
import strings.*;




class collection {
    //[1,2,3] -> (1; 2; 3)
    public static void main(String[] args) {
        List<Integer> list = Arrays.asList(1,2,3);
        System.out.println(list);
        System.out.println("-----------------------------");
        //3.2.1 각 문자열이 어떤 역할을 하는지 알 수 없다 -> 직접 함수 시크니쳐를 살펴봐야한다

        System.out.println(joinToString1(list, "; ", "(", ")"));
        System.out.println("-----------------------------");

        //3.2.2 동일한 메서드에 인자를 다르게 넘겨서 호출 할 경우가 생기는데 오버로딩 해서 다음과 같이 호출해줘야한다.
        System.out.println(joinToString(list)); // 매개변수 1개
        System.out.println(joinToString(list, "; ")); // 매개변수 2개
        System.out.println(joinToString(list, "; ", "(")); // 매개변수 3개
        System.out.println(joinToString(list, "; ", "(", ")")); // 매개변수 4개

        //3.2.3 유틸리티 함수로 직접 호출
        System.out.println(MathFunctions.add(1,2));
        System.out.println(StringFunctions.joinToString4(list, ";", "(", ")"));
    }


    /*
    3.2.1
     */
    public static <T> String joinToString1(
            Collection<T> collection,
            String separator,
            String prefix,
            String postfix) {

        StringBuilder result = new StringBuilder(prefix);

        int index = 0;
        for (T element : collection) {
            if (index > 0) result.append(separator);
            result.append(element);
            index++;
        }

        result.append(postfix);
        return result.toString();
    }


    /*
        3.2.2
        자바 코드를 보면 동일 메소드에 인자값을 다르게 넘겨줄때
        오버로딩해서 다음과 같이 처리를 해줘야하는 불편함이 있다.
    */
    public static <T> String joinToString(Collection<T> collection) {
        return joinToString(collection, ", ");
    }

    public static <T> String joinToString(Collection<T> collection, String separator) {
        return joinToString(collection, separator, "");
    }

    public static <T> String joinToString(Collection<T> collection, String separator, String prefix) {
        return joinToString(collection, separator, prefix, "");
    }

    public static <T> String joinToString(Collection<T> collection, String separator, String prefix, String postfix) {
        StringBuilder result = new StringBuilder(prefix);

        int index = 0;
        for (T element : collection) {
            if (index > 0) result.append(separator);
            result.append(element);
            index++;
        }

        result.append(postfix);
        return result.toString();
    }
}

/*
    특징: 구아바는 컬렉션, 캐싱, 프리미티브 지원, 동시성 라이브러리, 공통 어노테이션, 문자열 처리, I/O 등을 포함한 광범위한 유틸리티를 제공합니다.
    용도: 예를 들어, 구아바에서는 Joiner 클래스를 사용하여 컬렉션의 요소를 원하는 구분자로 연결할 수 있습니다. 이는 Java의 String.join 또는 java.util.stream.Collectors.joining과 유사한 기능을 제공하지만 더 많은 기능과 유연성을 가지고 있습니다.
    // 구아바를 사용한 문자열 합치기
    String joined = Joiner.on("; ").join(Arrays.asList(1, 2, 3));
    아파치 커먼즈 (Apache Commons)

    개발자: Apache Software Foundation
    특징: 아파치 커먼즈는 수학 연산, 파일 입출력, 문자열 처리, 네트워킹 등 다양한 목적의 유틸리티를 제공합니다.
    용도: 예를 들어, commons-lang 컴포넌트는 StringUtils 클래스를 제공하여 문자열 조작과 관련된 다양한 헬퍼 메서드를 사용할 수 있게 합니다.
    // 아파치 커먼즈를 사용한 문자열 체크
    boolean isNumeric = StringUtils.isNumeric("12345");
 */

File: kotlin-in-action-master/src/ch03/study/study_3_2/collection.kt
package ch03.study.study_3_2

import Maths.*
import strings.*
import strings.frist as fristNum


/* 3.2.1 */
fun <T> joinToString1(
    collection: Collection<T>,
    separator: String,
    prefix: String,
    postfix: String
): String {

    val result = StringBuilder(prefix)

    for ((index, element) in collection.withIndex()) {
        if (index > 0) result.append(separator)
        result.append(element)
    }

    result.append(postfix)
    return result.toString()
}


/*
    3.2.2
    디폴트 파라메터를 다름과 같이 명시해주면 인자값을 다르게 넘겨주기위해 동일한 메서드에 오버로딩 작업을 할 필요가 없다
*/
fun <T> joinToString2(
    collection: Collection<T>,
    separator: String = ", ",
    prefix: String= "",
    postfix: String = ""
): String {

    val result = StringBuilder(prefix)

    for ((index, element) in collection.withIndex()) {
        if (index > 0) result.append(separator)
        result.append(element)
    }

    result.append(postfix)
    return result.toString()
}

// [1,2,3] -> (1; 2; 3)
fun main(args: Array<String>) {

    val list = listOf(1,2,3);
    println(list);
    println("-----------------------------");

    //3.2.1 호출시 인자에 이름을 명시해서 간단하게 함수를 호출 가능하다 -> 각 문자열이 어떤 역할을 하는지 알수있다.
    println(joinToString1(list, separator=";", prefix="(", postfix=")"));
    println("-----------------------------");

    //3.2.2 호출시 오버로딩 (같은 이름의 메서드를 인자를 다르게 해서 여러번 정의) 할 필요없이 디폴트 파라메터값 명시로 이것을 해결
    println(joinToString2(list, separator=";", prefix="(", postfix=")"))
    println(joinToString2(list, separator=";", prefix="("));
    println(joinToString2(list, separator=";", postfix=")"));

    /* 3.2.3 유틸리티 함수로 호출 */
    println(add(1,2));
    println(joinToString4(list, separator=";", prefix="(", postfix=")"))

    /* 3.2.4 확장함수 & 임포트 함수 수정 */
    println("kotlin".fristNum());//수신객체 타입
    println("kotlin".test().fristNum());//확장함수
    println("kotlin".second());//확장함수
    println("Kotlin".fristNum());//임포트한 함수를 다른 이름으로 부를 수 있다.
}
